### ISSUE TITLE
numpy reduce operation on masked array with no mask fails (migrated from Trac #1166)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2719]
Original ticket http://projects.scipy.org/numpy/ticket/1166
Reported 2009-07-11 by trac user reggie, assigned to atmention:pierregm.

If you have a MaskedArray with no masked values, but whose mask is not nomask, a numpy operation on that array that results in an array whose shape is different from the original (eg. numpy.maximum.reduce), will result in an exception being raised.

For example:

```
In [1]: import numpy

In [2]: numpy.__version__
Out[2]: '1.3.0'

In [3]: x = numpy.ma.arange(5)

In [4]: x.mask = numpy.ma.getmaskarray(x)

In [5]: x
Out[5]: 
masked_array(data = [0 1 2 3 4],
             mask = [False False False False False],
       fill_value = 999999)


In [6]: numpy.maximum.reduce(x)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)

/home/reggie/src/python/<ipython console> in <module>()

/usr/local/lib/python2.6/site-packages/numpy/ma/core.py in __array_finalize__(self, obj)
   2444         # Finalize the mask ...........

   2445         if self._mask is not nomask:
-> 2446             self._mask.shape = self.shape
   2447         return
   2448     #..................................


ValueError: total size of new array must be unchanged

In [7]: 
```

I have attached a patch that seems to fix the problem.

**Comment by thouis** [comment]
Attachment in Trac by trac user reggie, 2009-07-11: [ma.patch](http://projects.scipy.org/numpy/attachment/ticket/1166/ma.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:pierregm, 2009-07-18

fixed in r7137.

Notes:
- np.maximum does not work on MaskedArray: use ma.maximum instead, whose .reduce method worked without modifications.
- I was not comfortable with the proposed patch. Using mask.any() systematically in **array_finalize** was likely to slow computation down. I chose to go with a try/except instead. However, I did not perform any actual optimization test.
